Draft of first half of banksy notebook (#984)
AlexsLemonade/training-modules: spatial/02-spatial_clustering.Rmd

---
title: "Spatially-aware clustering"
author: Data Lab for ALSF
date: 2026
output:
  html_notebook:
    toc: true
    toc_float: true
---

## Objectives

- Perform PCA, clustering, UMAP sans spatial information
- Use BANKSY to perform spatially-aware PCA and clustering
- Compare clusters calculated with and without spatial information

## Introduction

> We will be using a sample from a mouse with osteosarcoma lung metastases e from [Reinecke et al. (2025)](https://doi.org/10.1158/1078-0432.CCR-24-0049).

Reviewer: ```suggestion
We will be using a sample from a mouse with osteosarcoma lung metastases from [Reinecke et al. (2025)](https://doi.org/10.1158/1078-0432.CCR-24-0049).
```

suppressPackageStartupMessages({
  library(SpatialExperiment)
})

# To support plotting
library(ggspavis)
library(ggplot2)
library(patchwork)

# Set ggplot2 theme
theme_set(theme_bw())
```

Set a seed for reproducibility:

```{r}
set.seed(2026)
```

### Directories and files

```{r inputs}
# define sample id we'll be analyzing 
sample_id <- "GSM8478586"

# data directory for this sample
data_dir <- file.path("data/osteo", sample_id, "normalized")

normalized_rds <- file.path(data_dir, "osteo_normalized_spe.rds")
```

And the output file:

```{r}
clustered_rds <- file.path(data_dir, "osteo_clustered_spe.rds")
```

## Import

```{r import-rds, live = TRUE}
spe <- readr::read_rds(normalized_rds)
```

Look at it first:

```{r print-spe, live = TRUE}
# print the spe
spe
```

```{r plot-visium}
#| fig.width: 5

# define H&E plot for general use in notebook
he_plot <- plotVisium(spe, spots = FALSE, zoom = TRUE) +
  ggtitle("H&E") +
  # add styling to match other panels
  theme(
    plot.title = ggplot2::element_text(
      hjust = 0.5, 
      margin = ggplot2::margin(0, 0, 0.1, 0)
    ),
    panel.border = ggplot2::element_rect(color = "black", linewidth = 0.25)
  )

he_plot
```

> We can clearly tumors amid the lung parenchyma, which itself is more sparse because of the alveolar spaces which is just how lungs work.

Reviewer: ```suggestion
We can clearly see tumor cells amid the lung parenchyma, which itself is more sparse because of the alveolar spaces which is just how lungs work.
```

Let's get a sense of the RNA counts here by plotting some of the QC stats we calculated for you.
Note that these plots use right-censoring and set an upper limit on the color scale to be able to see lower value spots more clearly.

```{r show qc}
#| warning: false
#| message: false

> total_plot <- plotCoords(spe, annotate = "sum", point_size = 1.25) + 

Reviewer: I feel like the point size actually needs to be smaller here? The points are overlapping a lot which makes them harder to interpret.

  scale_color_viridis_c(
    # cap scale at 20,000; all value above this will be shown as 20,000
    limits = c(0, 20000), 
    oob = scales::oob_squish, 
    # fun with color scales!
    option = "plasma"
  )
detected_plot <- plotCoords(spe, annotate = "detected",  point_size = 1.25) + 
  scale_color_viridis_c(
    # cap scale at 10,000
    limits = c(0, 10000), 
    oob = scales::oob_squish, 
    option = "plasma"
  )
total_plot + detected_plot
```
Takeaways:

- Clearly, more expression in the tumors in general vs parenchyma, which makes sense biologically!

> - Because we have really different tissue compartments here, if we had used global thresholds we would have just ended up retaining almost exclusively tumor cells and removing their spatial context due to lower expression, but we would like to retain this spatial context because otherwise we've just wasted a ton of money on Visium

Reviewer: 💯

- Finally we can cluster/run UMAP on this PCA

See more details about parameters [from these docs](https://prabhakarlab.github.io/Banksy/articles/parameter-selection.html#parameters).

We're going to run this without and with AGF and then compare all around.

#### Without AGF

```{r banksy matrix agf off, live = TRUE}
# define parameters recommended for Visium
lambda <- 0.2
k_geom <- 18

# calcuate the BANKSY matrix without AGF
spe <- Banksy::computeBanksy(spe, assay_name = "logcounts", k_geom = k_geom)

# show spe
spe
```

Show and explain the H0 matrix (spatially-smoothed neighbor-averaged expression).

```{r show ho}
assay(spe, "H0")[1:5, 1:5]
```
Now we can calculate PCA, specifying `lambda` for how to weight the original expression (`logcounts`) and the BANKSY matrix (here, H0, 

```{r banksy pca, live = TRUE}
# calculate PCA within BANKSY
spe <- Banksy::runBanksyPCA(spe, lambda = lambda)

# show SPE
spe
```
We have a new PCA called `PCA_M0_lam0.2`.
BANKSY named this for us: `M0` implies that AGF is off, and `lam0.2` is our lambda parameter.

We can proceed to calculate UMAP and clusters now, and we'll use `scran`/`scater` functions for that.

> But, it's worth point out that BANKSY does allow you to specify multiple `lambda` parameters at once and then offers wrapper functions to cluster/UMAP across those parameters in one function call.

Reviewer: ```suggestion
But, it's worth pointing out that BANKSY does allow you to specify multiple `lambda` parameters at once and then offers wrapper functions to cluster/UMAP across those parameters in one function call.
```
